WordPress/gutenberg
# Env: Download WordPress PHPUnit Into Container

## What?

This pull request downloads the WordPress PHPUnit test files into the `wordpress` and `tests-wordpress` container. This closes #23171.

## Why?

Currently, the best option we have for installing the WordPress PHPUnit test library is using `wp-phpunit/wp-phpunit`. This is not a _bad_ solution, however, it comes with a few problems:

* If you select a version of WordPress that is incompatible with the installed version of `wp-phpunit/wp-phpunit` your tests will fail.
* There is no option to download the current _trunk_ test files. This can lead to problems when trying to test against the next version of WordPress.

Since we know what version of WordPress the environment contains, a better solution would be to just download it and mount it in the container.

## How?

This pull request works by scanning `~/.wp-env/{hash}/{coreSource}/wp-includes/version.php` for the current WordPress version. An alternative that was considered was using `wp core version`, but, I wanted to find a solution that could give us the version to include the library in the docker-compose file.

We can then use this information to figure out what version of the WordPress PHPUnit library to install. We store it in the `~/.wp-env/{hash}/` directory and mount it at `/phpunit` in the container. To get around the need for anything other than `git`, we download from [WordPress/wordpress-develop](https://github.com/WordPress/wordpress-develop). It mirrors from SVN so we can rely on `master` being the bleeding edge and the tags containing WordPress versions. We're also using a sparse checkout to prevent having the _entire_ `WordPress/wordpress-develop` repository checked out.

Lastly, in order to use it in unit tests, we copy the `wp-config.php` file and set the required constants. Once we create the containers with the `WP_PHPUNIT__DIR` we can now run unit tests without `wp-phpunit/wp-phpunit`

**Note: I looked at removing `wp-phpunit/wp-phpunit` but it felt out of scope for this PR. It requires changes to the test suite to properly use `Yoast/PHPUnit-Polyfills`.**

## Testing Instructions

This is tricky to test, unfortunately. We're going to have to remove `wp-phpunit/wp-phpunit`. The tests won't pass unless you do the legwork, but, just removing it can let you see that the test suite boots up. That is really all you need to know this in particular works.

1. Run `composer remove wp-phpunit/wp-phpunit --dev` to get rid of `wp-phpunit/wp-phpunit`. In theory, the PHPUnit tests should now not even be able to boot.
2. Run `npm run wp-env run tests-wordpress "/var/www/html/wp-content/plugins/gutenberg/vendor/bin/phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist --verbose"`
3. Note that the PHPUnit suite boots up and loads WordPress but fails because of `wp-phpunit/wp-phpunit` failures.
4. Poke around `~/.wp-env/{hash}` and check out the `WordPress-PHPUnit` and `tests-WordPress-PHPUnit` directories. They should contain all of the test files with a sparse checkout.
5. Use the `WP_ENV_CORE` environment variable to try out a few different options for WordPress installs. A local one, a zip, a git repository, etc. Verify that the correct version of the PHPUnit library is installed in each case.

I have prepared https://github.com/WordPress/gutenberg/pull/41852 with all of the commits from this pull request. It makes the appropriate changes for the Gutenberg PHPUnit tests to be ran in `wp-env`. The PHPUnit test workflow in that pull request uses the provided test files. That pull request passing is a demonstration of this working in CI.
